pwndbg/lib/memoize.py

```diff
@@ -12,173 +12,83 @@
-        return "<%s-memoized function %s>" % (self.kind, funcname)
-
-    def __get__(self, obj, objtype: type) -> Callable:
-        return functools.partial(self.__call__, obj)
-
-    def clear(self) -> None:
-        if debug:
-            print("Clearing %s %r" % (self, self.cache))
-        self.cache.clear()
-
-
-class forever(memoize):
-    """
-    Memoizes forever - for a pwndbg session or until `_reset` is called explicitly.
-    """
-    caches = []  # type: List[forever]
-    kind = "forever"
-
-    @staticmethod
-    def _reset() -> None:
-        for obj in forever.caches:
-            obj.cache.clear()
-
-
-class reset_on_stop(memoize):
-    caches = []  # type: List[reset_on_stop]
-    kind = "stop"
-
-    @staticmethod
-    def __reset_on_stop() -> None:
-        for obj in reset_on_stop.caches:
-            obj.cache.clear()
-
-    _reset = __reset_on_stop
-
-
-class reset_on_prompt(memoize):
-    caches = []  # type: List[reset_on_prompt]
-    kind = "prompt"
-
-    @staticmethod
-    def __reset_on_prompt() -> None:
-        for obj in reset_on_prompt.caches:
-            obj.cache.clear()
-
-    _reset = __reset_on_prompt
-
-
-class reset_on_exit(memoize):
-    caches = []  # type: List[reset_on_exit]
-    kind = "exit"
-
-    @staticmethod
-    def __reset_on_exit() -> None:
-        for obj in reset_on_exit.caches:
-            obj.clear()
-
-    _reset = __reset_on_exit
-
-
-class reset_on_objfile(memoize):
-    caches = []  # type: List[reset_on_objfile]
-    kind = "objfile"
-
-    @staticmethod
-    def __reset_on_objfile() -> None:
-        for obj in reset_on_objfile.caches:
-            obj.clear()
-
-    _reset = __reset_on_objfile
-
-
-class reset_on_start(memoize):
-    caches = []  # type: List[reset_on_start]
-    kind = "start"
-
-    @staticmethod
-    def __reset_on_start() -> None:
-        for obj in reset_on_start.caches:
-            obj.clear()
-
-    _reset = __reset_on_start
-
-
-class reset_on_cont(memoize):
-    caches = []  # type: List[reset_on_cont]
-    kind = "cont"
-
-    @staticmethod
-    def __reset_on_cont() -> None:
-        for obj in reset_on_cont.caches:
-            obj.clear()
-
-    _reset = __reset_on_cont
-
-
-class reset_on_thread(memoize):
-    caches = []  # type: List[reset_on_thread]
-    kind = "thread"
-
-    @staticmethod
-    def __reset_on_thread() -> None:
-        for obj in reset_on_thread.caches:
-            obj.clear()
-
-    _reset = __reset_on_thread
-
-
-class while_running(memoize):
-    caches = []  # type: List[while_running]
-    kind = "running"
-    caching = False
-
-    @staticmethod
-    def _start_caching() -> None:
-        while_running.caching = True
-
-    @staticmethod
-    def __reset_while_running() -> None:
-        for obj in while_running.caches:
-            obj.clear()
-        while_running.caching = False
-
-    _reset = __reset_while_running
+debug = 0
+
+# function -> function wrapper
+wrappers = {}
```

Comment:
Would appreciate some type hints for globals like this (doesn't need to be perfect, even `Dict[Callable, Callable]` is fine if it's not easy to be more accurate).